src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -100,12 +103,33 @@ public void storeMessage(Message message) {
     // people may include implicit context/reference to previously sent messages,
     // and this order makes the chronology of the conversation and context
     // more intuitive and understandable in the messages list page. 
-    Query query =
-        new Query("Message")
-            .setFilter(directMessages)
-            .addSort("timestamp", SortDirection.ASCENDING);
+    Query query = hasUserRestriction(loggedInUser, otherUser)
+                        ? createMessageQueryWithUserFilter(loggedInUser, otherUser, directMessages)
+                        : createPublicMessageQuery();
+    
     PreparedQuery results = datastore.prepare(query);
 
+    return entityToMessage(messages, results);
+  }
+
+  /** Returns the total number of messages for all users. */
+  public int getTotalMessageCount() {
+    Query query = new Query("Message");
+    PreparedQuery results = datastore.prepare(query);
+    return results.countEntities(FetchOptions.Builder.withDefaults());
+  }
+  public List<Message> getAllMessages() {
+    List<Message> messages = new ArrayList<>();
+    
+    Query query = 
+        new Query("Message")
+          .addSort("timestamp", SortDirection.ASCENDING);
+    
+    PreparedQuery results = datastore.prepare(query);
+    
+    return entityToMessage(messages, results);
+  }
+  private List<Message> entityToMessage(List<Message> messages, PreparedQuery results) {
```

Comment:
Minor feedback: `entityToMessage` implies a single `Entity` is being converted to a single `Message`.  A method like this should probably be named `preparedQueryToMessages`. It is, in my view, confusing that data is returned through an output parameter. If you have an output parameter, it would be better to return `void`. However, even better would be to return data through the `return` statement only, and not via output parameters.

My recommendation to you:

```
  private List<Message> preparedQueryToMessages(PreparedQuery query) {
     List<Message> results  = new ArrayList<>();
     for (Entity entity : query.asIterable()) {
        results.add(entityToMessage(entity));
     }
     return results;
  }

  private Message entityToMessage(Entity entity) {
      String idString = entity.getKey().getName();
      UUID id = UUID.fromString(idString);
      String text = getStringProperty(entity, "text").orElse("");
      long timestamp = (long) entity.getProperty("timestamp");
      String sender = (String) entity.getProperty("user");
      String receiver = (String) entity.getProperty("recipient");
      return new Message(id, sender, text, timestamp, receiver);
  }
```

Reply:
Yeah, this makes more sense than the spaghetti code I wrote.
Thank you for the feedback!